pkg/statedb/db.go

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package statedb
+
+import (
+	"encoding/json"
+	"fmt"
+
+	memdb "github.com/hashicorp/go-memdb"
+
+	"github.com/cilium/cilium/pkg/stream"
+)
+
+func New(p params) (DB, error) {
+	dbSchema := &memdb.DBSchema{
+		Tables: make(map[string]*memdb.TableSchema),
+	}
+	for _, tableSchema := range p.Schemas {
+		if _, ok := dbSchema.Tables[tableSchema.Name]; ok {
+			panic(fmt.Sprintf("Table %q already registered", tableSchema.Name))
+		}
+		dbSchema.Tables[tableSchema.Name] = tableSchema
+	}
+	memdb, err := memdb.NewMemDB(dbSchema)
+	if err != nil {
+		return nil, err
+	}
+	db := &stateDB{
+		memDB:    memdb,
+		revision: 0,
+	}
+	db.Observable, db.emit, _ = stream.Multicast[Event]()
+
+	return db, nil
+}
+
+// stateDB implements StateDB using go-memdb.
+type stateDB struct {
+	stream.Observable[Event]
+	emit func(Event)
+
+	memDB    *memdb.MemDB
+	revision uint64 // Commit revision, protected by the write tx lock.
+}
+
+var _ DB = &stateDB{}
+
+// ToJSON marshals the whole database into a JSON string.
+func (db *stateDB) ToJSON() ([]byte, error) {
+	tx := db.memDB.Txn(false)
+	out := map[string][]any{}
+	for table := range db.memDB.DBSchema().Tables {
+		iter, err := tx.Get(table, "id")
+		if err != nil {
+			panic(err)
```

Comment:
is this really panic-worthy?

Reply:
It'll only fail if the table schema is bad (e.g. the mandatory "id" index not defined or faulty). But it doesn't seem like this method should be the one panicing so I rewrote this to return the error and to only marshal one object at a time into a io.Writer rather than all of them in one go.